addons/sourcemod/scripting/superzombiefortress.sp

```diff
@@ -979,55 +981,129 @@ public Action hook_JoinTeam(int client, const char[] command, int argc)
+	
+	//Check if client is trying to skip playing as zombie by joining spectator
+	if (StrEqual(cmd1, "spectate", false))
+		CheckZombieBypass(client);
+			
+	// Assign team-specific strings
+	if (zomTeam() == INT(TFTeam_Blue))
+	{
+		sSurTeam = "red";
+		sZomTeam = "blue";
+		sZomVgui = "class_blue";
+	}
+	else
+	{
+		sSurTeam = "blue";
+		sZomTeam = "red";
+		sZomVgui = "class_red";
+	}
 
-	if (roundState() >= RoundGrace)
+	if (IsClientInGame(client))
 	{
-		// Assign team-specific strings
-		if (zomTeam() == INT(TFTeam_Blue))
+		if (roundState() == RoundGrace)
 		{
-			sSurTeam = "red";
-			sZomTeam = "blue";
-			sZomVgui = "class_blue";
-		}
-		else
-		{
-			sSurTeam = "blue";
-			sZomTeam = "red";
-			sZomVgui = "class_red";
-		}
-		
-		//Check if client is trying to skip playing as zombie by joining spectator
-		if (StrEqual(cmd1, "spectate", false))
-			CheckZombieBypass(client);
+			int iTeam = GetClientTeam(client);
 		
-		// If client tries to join the survivor team or a random team
-		// during grace period or active round, place them on the zombie
-		// team and present them with the zombie class select screen.
-		if (StrEqual(cmd1, sSurTeam, false) || StrEqual(cmd1, "auto", false))
-		{
-			ChangeClientTeam(client, zomTeam());
-			ShowVGUIPanel(client, sZomVgui);
-			return Plugin_Handled;
-		}
-		// If client tries to join the zombie team or spectator
-		// during grace period or active round, let them do so.
-		else if (StrEqual(cmd1, sZomTeam, false) || StrEqual(cmd1, "spectate", false))
-		{
-			return Plugin_Continue;
+			// If a client tries to join the infected team or a random team during grace period...
+			if (StrEqual(cmd1, sZomTeam, false) || StrEqual(cmd1, "auto", false) || StrEqual(cmd1, "autoteam", false))
+			{
+				// ...as survivor, don't let them.
+				if (iTeam == surTeam())
+				{
+					CPrintToChat(client, "{red}You can not switch to the opposing team during grace period.");
+					return Plugin_Handled;
+				}
+
+				// ...as a spectator who didn't start as an infected, set them as infected after grace period ends, after warning them.
+				if (iTeam <= 1 && !g_bStartedAsZombie[client])
+				{
+					if (!g_bWaitingForTeamSwitch[client])
+					{
+						if (iTeam == INT(TFTeam_Unassigned)) // If they're unassigned, let them spectate for now.
+							ChangeClientTeam(client, INT(TFTeam_Spectator));
+							
+						CPrintToChat(client, "{red}You will join the Infected team when grace period ends.");
+						g_bWaitingForTeamSwitch[client] = true;
+					}
+					return Plugin_Handled;
+				}	
+			}
+
+			// If client tries to spectate during grace period, make them
+			// not be booted into the infected team if they tried to join
+			// it before.
+			else if (StrEqual(cmd1, "spectate", false))
+			{
+				if (iTeam <= 1 && g_bWaitingForTeamSwitch[client])
+				{					
+					CPrintToChat(client, "{red}You will no longer automatically join the Infected team when grace period ends.");
```

Comment:
Would fit better for green color rather than red, as this situation isn't a bad thing